src/EventHandler/Fun/Pipe/PipeArgumentsProvider.php

```diff
@@ -4,275 +4,94 @@
-use Psalm\Plugin\EventHandler\Event\FunctionReturnTypeProviderEvent;
-use Psalm\Plugin\EventHandler\FunctionParamsProviderInterface;
-use Psalm\Plugin\EventHandler\FunctionReturnTypeProviderInterface;
-use Psalm\StatementsSource;
+use Psalm\Plugin\DynamicFunctionStorage;
+use Psalm\Plugin\DynamicTemplateProvider;
+use Psalm\Plugin\EventHandler\DynamicFunctionStorageProviderInterface;
+use Psalm\Plugin\EventHandler\Event\DynamicFunctionStorageProviderEvent;
 use Psalm\Storage\FunctionLikeParameter;
-use Psalm\Type;
+use Psalm\Type\Atomic\TCallable;
+use Psalm\Type\Union;
 
-/**
- * @psalm-type Stage = array{0: Type\Union, 1: Type\Union, 2: string}
- * @psalm-type StagesOrEmpty = list<Stage>
- * @psalm-type Stages = non-empty-list<Stage>
- */
-class PipeArgumentsProvider implements FunctionParamsProviderInterface, FunctionReturnTypeProviderInterface
+use function array_map;
+use function count;
+use function range;
+
+final class PipeArgumentsProvider implements DynamicFunctionStorageProviderInterface
 {
     /**
      * @return array<lowercase-string>
      */
     public static function getFunctionIds(): array
     {
-        return [
-            'psl\fun\pipe'
-        ];
+        return ['psl\fun\pipe'];
     }
 
-    /**
-     * @return list<FunctionLikeParameter>|null
-     */
-    public static function getFunctionParams(FunctionParamsProviderEvent $event): ?array
+    public static function getFunctionStorage(DynamicFunctionStorageProviderEvent $event): ?DynamicFunctionStorage
     {
-        $stages = self::parseStages($event->getStatementsSource(), $event->getCallArgs());
-        if (!$stages) {
-            return [];
-        }
-
-        $params = [];
-        $previousOut = self::pipeInputType($stages);
-
-        foreach ($stages as $stage) {
-            [$_, $currentOut, $paramName] = $stage;
-
-            $params[] = self::createFunctionParameter(
-                'stages',
-                self::createClosureStage($previousOut, $currentOut, $paramName)
-            );
-
-            $previousOut = $currentOut;
-        }
-
-        return $params;
-    }
-
-    public static function getFunctionReturnType(FunctionReturnTypeProviderEvent $event): ?Type\Union
-    {
-        $stages = self::parseStages($event->getStatementsSource(), $event->getCallArgs());
-        if (!$stages) {
-            //
-            // @see https://github.com/vimeo/psalm/issues/7244
-            // Currently, templated arguments are not being resolved in closures / callables
-            // For now, we fall back to the built-in types.
-
-            // $templated = self::createTemplatedType('T', Type::getMixed(), 'fn-'.$event->getFunctionId());
-            // return self::createClosureStage($templated, $templated, 'input');
+        $template_provider = $event->getTemplateProvider();
+        $callable_args_count = count($event->getArgs()) - 1;
 
+        if ($callable_args_count < 1) {
             return null;
```

Comment:
wip here, I copied a plugin from the author of the new dynamic storage.
I'm gonna look if I can add support for lesser arguments as well